src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -311,8 +311,11 @@ boost::optional<std::string> PeerReplayer::pick_directory() {
       break;
     }
   }
-
-  std::rotate(m_directories.begin(), m_directories.begin() + 1, m_directories.end());
+  if (m_directories.size() > 1) {
```

Comment:
Isn't pick_directory called only when m_directories is not empty?

Reply:
If there is on element in `pick_directory` the `m_directories.begin() + 1` will be `.end()`. Is that okay ?

Reply:
yeah, i checked the code venky, you're right, it calls `pick_directory()` only after checking size of `m_directories`.

Reply:
we have `auto &sync_stat = m_snap_sync_stats.at(dir_root);` in `pick_directory()` code, if `dir_root` is not present in the map then it might throw error
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
Aborted (core dumped)
```

Reply:
> we have `auto &sync_stat = m_snap_sync_stats.at(dir_root);` in `pick_directory()` code, if `dir_root` is not present in the map then it might throw error
> 
> ```
> terminate called after throwing an instance of 'std::out_of_range'
>   what():  map::at
> Aborted (core dumped)
> ```
> 
> i see dirs addition in `m_snap_sync_stats` being done in `add_directory()` and `init()` while deletion in `remove_directory()` and `unregister_directory()`, chances can be maybe in a scenario, `pick_directory()` was called for a dir that was removed? or some thread ran it before the addition?

These operations are done safely under a lock, isn't it?

Reply:
hmm, i just checked, all these methods are called after acquiring a lock

Reply:
i can say there is no issue in `pick_directory()` then

Reply:
Yeh. Looks unlikely then. The tracker is for v17.2.1 and things might have changed b/w minor versions.

Reply:
looks so, i'll try to probe the code again today, let's see if i can find something else i would it be good to close this for now with the tracker marked as needinfo?

Reply:
If you are closing the tracker, use "can't reproduce" in the tracker status.